### PR TITLE
[ML] Store and restore training jobs with  models

### PR DESCRIPTION
**Pull request by valeriy42**
This PR adds functionality to store and restore training jobs. It mainly requires storing the model blob object, but other metadata like `input_filename`, `dependent_variable`, `analysis_name` are also stored for completeness.  Jobs are stored as gzipped JSON files.

The job can be stored with 
```python
job.store(destination=local_path)
```
and restored with 
```python
job = Job.fromFile(source=local_path)
```
See unit tests in `test_store_restore.py` for more examples.

Furthermore, I implemented functionality to upload, download and delete job files in the Google storage bucket.

### CODE REVIEW

File: jupyter/src/incremental_learning/job.py [hunk 1]
```diff
@@ -272,6 +280,101 @@ def get_model_update_data(self) -> str:
...
+            try:
+                json.dump(obj=self, fp=fp, cls=JobJSONEncoder)
+            except:
+                ex = sys.exc_info()
+                logger.error("Failed storing job {}: {}".format(self.name, ex))
+                success = False
+        return success
+
+    @classmethod
+    def asJob(cls, state: dict):
```
Comment (tveasey) [inline jupyter/src/incremental_learning/job.py:304]: nit: stick with snake case conventions used in this file
```suggestion
    def as_job(cls, state: dict):
```

File: jupyter/src/incremental_learning/job.py [hunk 2]
```diff
@@ -272,6 +280,101 @@ def get_model_update_data(self) -> str:
...
+            job.restore_filename = state['restore_filename']
+            job.name = state['name']
+            job.model = state['model']
+            if 'results' in state:
+                job.results =  state['results']
+            job.initialized = True
+        return job
+
+    @classmethod
+    def fromFile(cls, source: Path):
```
Comment (tveasey) [inline jupyter/src/incremental_learning/job.py:322]: As before
```suggestion
    def from_file(cls, source: Path):
```

**Review by tveasey — APPROVED** [review]
Nice! Functionality looks good; left a couple of very minor nits on naming, otherwise LGTM.

File: jupyter/src/incremental_learning/job.py (hunk 1 above)
Comment (valeriy42) [reply jupyter/src/incremental_learning/job.py:304]: renamed

File: jupyter/src/incremental_learning/job.py (hunk 2 above)
Comment (valeriy42) [reply jupyter/src/incremental_learning/job.py:322]: renamed

**Pull request merged**